src/ipa_pc/mod.rs

```diff
@@ -378,106 +378,126 @@ impl<G: AffineCurve, D: Digest> InnerProductArgPC<G, D> {
+        let mut cur_challenge = G::ScalarField::one();
+
+        // Fresh random challenge x
+        fs_rng.absorb(&to_bytes![batch_proof.h_comm].unwrap());
+        let x_point: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
 
-        // y_i values
-        let mut y_values = vec![];
+        // LC(C): reconstructed commitment to LC(p_1(X),p_2(X),...,p_m(X),h(X))
+        let mut lc_comm = vec![];
 
-        // x_i values
-        let mut points = vec![];
+        // Expected value wich LC(p_1(X),p_2(X),...,p_m(X),h(X)) opens to
+        let mut lc_value = G::ScalarField::zero();
 
         for (label, (_point_label, point)) in query_set.iter() {
+            let labeled_commitment =
+                *commitment_map.get(label).ok_or(Error::MissingCommitment {
+                    label: label.to_string(),
+                })?;
 
-            let y_i = values
+            // y_i
+            let y_i = *values
                 .get(&(label.clone(), *point))
                 .ok_or(Error::MissingEvaluation {
                     label: label.to_string(),
                 })?;
-            y_values.push(*y_i);
-
-            let v_i = batch_proof.batch_values[
-                *labels
-                    .get(&label.clone())
-                    .ok_or(Error::MissingEvaluation {
-                        label: label.to_string(),
-                    })?
-            ];
-            v_values.push(v_i);
-
-            points.push(point);
-        }
-
-        // Commitment of the h(X) polynomial
-        let batch_commitment = batch_proof.batch_commitment.clone();
 
-        // lambda
-        let lambda: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
-        let mut cur_challenge = G::ScalarField::one();
+            // (X - x_i)
+            let x_polynomial = Polynomial::from_coefficients_slice(&[
+                -(*point),
+                G::ScalarField::one()
+            ]);
 
-        // Fresh random challenge x
-        fs_rng.absorb(&to_bytes![batch_commitment].unwrap());
-        let point: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
+            // z_i(X) = z(X) / (X - x_i)
+            let z_i_polynomial = &z_polynomial / &x_polynomial;
 
-        let mut computed_batch_v = G::ScalarField::zero();
+            // z_i(x)
+            let z_i_value = z_i_polynomial.evaluate(x_point);
 
-        for ((&v_i, y_i), x_i) in v_values.iter().zip(y_values).zip(points) {
+            // LC(C) = SUM ( lambda^i * z_i(x) * C_i )
+            for (index, &comm_dash_segment) in labeled_commitment.commitment().comm.iter().enumerate() {
+                if index == lc_comm.len() {
+                    lc_comm.push(G::zero());
+                }
+                lc_comm[index] = (lc_comm[index].into_projective()
+                    + &(comm_dash_segment.into_projective()
+                        .mul(&z_i_value)
+                        .mul(&cur_challenge)
```

Comment:
```suggestion
                        .mul(&(z_i_value * &cur_challenge))
```

Reply:
Right.  I missed that too. Thanks to our large scale benches for the batch verifier we found it.